bumble/device.py

```diff
@@ -1536,6 +1547,135 @@ async def stop_advertising(self) -> None:
+        target: Address = Address.ANY,
+        own_address_type: int = OwnAddressType.RANDOM,
+        scan_response: Optional[bytes] = None,
+        advertising_data: Optional[bytes] = None,
+    ) -> int:
+        """Starts an extended advertising set.
+
+        Args:
+          advertising_properties: Properties to pass in HCI_LE_Set_Extended_Advertising_Parameters_Command
+          target: Directed advertising target. Directed property should be set in advertising_properties arg.
+          own_address_type: own address type to use in the advertising.
+          scan_response: raw scan response. When a non-none value is set, HCI_LE_Set_Extended_Scan_Response_Data_Command will be sent.
+          advertising_data: raw advertising data. When a non-none value is set, HCI_LE_Set_Advertising_Set_Random_Address_Command will be sent.
+
+        Returns:
+          Handle of the new advertising set.
+        """
+
+        adv_handle = -1
+        # Find a free handle
+        for i in range(
+            DEVICE_MIN_EXTENDED_ADVERTISING_SET_HANDLE,
+            DEVICE_MAX_EXTENDED_ADVERTISING_SET_HANDLE + 1,
+        ):
+            if i not in self.extended_advertising_handles:
+                adv_handle = i
+                break
+
+        if adv_handle == -1:
+            raise InvalidStateError('No available advertising set.')
+        self.extended_advertising_handles.add(adv_handle)
+
+        # Set the advertising parameters
+        await self.send_command(
+            HCI_LE_Set_Extended_Advertising_Parameters_Command(
+                advertising_handle=adv_handle,
+                advertising_event_properties=advertising_properties,
+                primary_advertising_interval_min=self.advertising_interval_min,
+                primary_advertising_interval_max=self.advertising_interval_max,
+                primary_advertising_channel_map=(
+                    HCI_LE_Set_Extended_Advertising_Parameters_Command.ChannelMap.CHANNEL_37
+                    | HCI_LE_Set_Extended_Advertising_Parameters_Command.ChannelMap.CHANNEL_38
+                    | HCI_LE_Set_Extended_Advertising_Parameters_Command.ChannelMap.CHANNEL_39
+                ),
+                own_address_type=own_address_type,
+                peer_address_type=target.address_type,
+                peer_address=target,
+                advertising_tx_power=7,
+                advertising_filter_policy=0,
+                primary_advertising_phy=1,  # LE 1M
+                secondary_advertising_max_skip=0,
+                secondary_advertising_phy=1,  # LE 1M
+                advertising_sid=0,
+                scan_request_notification_enable=0,
+            ),  # type: ignore[call-arg]
+            check_result=True,
+        )
+
+        # Set the advertising data if present
+        if advertising_data is not None:
+            await self.send_command(
```

Comment:
If this command is rejected (invalid parameter, if for example the data is too big, or if there are not enough advertising sets allowed in the controller), this will leak the handle. 
The same issue may happen with the following two commands (they could fail, and in that case not only would the handle leak, but also a partial state may be set in the controller)
It would be good to have a way to clean up any intermediate state if any command fails here, before raising an error to the caller.

Reply:
Nice catch. It also reminds me that we have to remove the advertising set on stop.